keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,302 @@
+
+- Scheduler does not put pods that require CSI volumes to nodes that don't have the CSI driver installed.
+
+- Scheduler does not increase its CPU consumption. Any regression must be approved by sig-scheduling.
+  - Scheduler benchmark must be extended to schedule pods with volumes as part of this KEP.
+
+### Non-Goals
+
+- Heterogenous clusters, i.e. clusters where access to storage is limited only to some nodes. Existing `PV.spec.nodeAffinity` handling, not modified by this KEP, will filter out nodes that don't have access to the storage, so predicates changed in this KEP don't need to worry about storage topology and can be simpler.
+
+- Multiple plugins sharing the same volume limits. We expect that every CSI driver will have its own limits, not shared with other CSI drivers. In this KEP we support only in-tree volume plugins sharing its limits with one hard-coded CSI driver each.
+
+- Multiple "units" per single volume. Each volume used on a node takes exactly 1 unit from `allocatable.volumes`, regardless of the volume size, its replica count, number of connections to remote servers or other underlying resources needed to use the volume. For example, multipath iSCSI volume with three paths (and thus three iSCSI connections to three different servers) still takes 1 unit from `CSINode.spec.drivers[xyz].allocatable.volumes`.
+
+- Maximum capacity per node. Some cloud environments limit both number of attached volumes (covered in this KEP) and total capacity of attached volumes (not covered in this KEP). For example, this KEP will ensure that scheduler puts max. 128 GCE PD volumes to a [typical GCE node](https://cloud.google.com/compute/docs/machine-types#predefined_machine_types), but it won't ensure that the total capacity of the volumes is less than 64 TB.
+
+## Proposal
+
+* Track volume limits for both in-tree volume plugins and CSI drivers in `CSINode` objects instead of `Node`.
+  * To get rid of prefix + SHA for `ResourceName` of CSI volumes.
+  * So in-tree volume plugin can share limits with CSI driver that uses the same storage backend.
+
+* Kubelet will create `CSINode` instance during initial node registration together with `Node` object.
+  * Limits of each in-tree volume plugin will be added to `CSINode.spec.drivers[xyz].allocatable`.
+    * Limit for in-tree volumes will be added by kubelet during CSINode creation. Name of corresponding CSI driver will be used as key in `CSINode.status.allocatable` and it will be discovered using [CSI translation library](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/csi-translation-lib). If the library does not support migration of an in-tree volume plugin, the volume plugin has no limit.
+    * TODO: this means we can't differentiate kubelet with CSI driver installed and not-installed for in-tree volumes! External provisioner may miss e.g. topology keys!
```

Comment:
Now kubelet creates e.g. `CSINode.spec.drivers[aws].allocatable` with limits for in-tree volumes during node registration. There is no AWS CSI driver registered yet, but it has entry in `CSINode.spec`.  Can it break anything?

* Scheduler will put pods with CSI volumes to the node. Since there is limit from in-tree plugin, nothing really bad can happen. When CSI driver is installed later, it may have different limits and the node may have too many volumes from the driver perspective, but it should not be dramatically over the quota - in-tree volume plugins already set the limits correctly.

* External provisioner does not see `TopologyKeys` there and may provision a wrong volume.

Reply:
Having nil/empty fields for nodeid and topology keys was awkward and difficult to explain api-wise when we had it this way for initial migration API design. Prefer to avoid it if possible.  

Can in-tree just continue to only register on node object? If migration is turned on, then it will be in csinode annotation and scheduler can use that to switch over to waiting for driver registration.

FYI, external provisioner [fails](https://github.com/kubernetes-csi/external-provisioner/blob/master/pkg/controller/topology.go#L309) if topology keys not found

Reply:
It would be much easier if there was single source of volume limits for all volume plugins. It would be IMO very confusing to users if the limits are in Node and suddenly move to CSINode and back as a driver is registered / deregistered.

Reply:
What I'm suggesting is:

* Migration is off: only use Node.  in-tree and csi will be counted differently and might conflict 
* Migration is on (for that plugin): only use CSINode

There's no switching between CSINode and Node objects when a driver gets installed.

Reply:
Checking if migration is on or not is non-trivial. For one thing - it requires plumbing volume plugin manager interface into scheduler which has been a no-no. 

But I think having entries in `CSINode.Spec.Drivers` might have other problems too. I was looking at how we are using inside operation_generator and it appears that if driver contains entry for a driver which is not installed, code will expect it to use CSI driver - https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_generator.go#L1773 and any operation will pretty much fail in unexpected ways (rather than just failing clearly with driver not installed yet).

Reply:
As part of @davidz627 's migration changes, he is adding an annotation to the CSINode object with the list of drivers where migration is on.  And the CSINode object needs to be created before the Node becomes ready.

Reply:
It is true that with the lifecycle of the API object we switched back to the existence of the Driver in the Spec is equivilant to the driver being `Installed` since we no longer have an explicit `Installed` field.
Additionally yeah, you can tell whether a driver is Migrated on a particular node by looking at the annotation on the CSINode object.

Reply:
Okay I have updated the spec to *not* create `CSINode.Spec.Drivers` entries for all in-tree volume types by default. Also, I have updated the spec to reflect that, for a in-tree volume type if migration is enabled but `CSINode.Spec.Drivers` don't have an entry yet then existing `Max***` predicates  will defer to CSI predicate for final decision.  

Since traditionally CSI volume predicate did not stop pods from scheduling a volume if driver is missing, we will have to follow some deprecation cycle before by default rejecting the pods.

Reply:
Another thing worth considering is - we can register limit for in-tree drivers with their in-tree names in `CSINode` object. This will ensure that, a CSI driver is not erroneously registered if CSI driver is not installed.

Reply:
cc @davidz627 @verult

Reply:
I don't think we should use the in-tree names since this will then diverge the in-tree with their migrated/csi counterparts which is what we're trying to avoid with this design. Additionally, the CSI Driver should never be registered if the CSI Driver is not installed. I looked at the 4 cases of migration/driver installed in the design currently and it looks like that is the case

Reply:
No - I was just proposing that we could create Driver entries for in-tree driver even if CSI driver is not installed.  The current document as it exists - does not create Driver entries in `CSINode` object if CSI driver is not installed.